api/src/main/java/io/minio/MinioClient.java

```diff
@@ -993,14 +999,40 @@ private boolean shouldOmitPortInHostHeader(HttpUrl url) {
    * @param length         Length of HTTP request body.
    */
   private HttpResponse execute(Method method, String region, String bucketName, String objectName,
-                               Map<String,String> headerMap, Map<String,String> queryParamMap,
+                               Map<String,String> simpleHeaderMap, Map<String,String> simpleQueryParamMap,
+                               Object body, int length)
+    throws InvalidBucketNameException, NoSuchAlgorithmException, InsufficientDataException, IOException,
+           InvalidKeyException, NoResponseException, XmlPullParserException, ErrorResponseException,
+           InternalException {
+
+    Multimap<String, String> queryParamMap = null;
+    if (simpleQueryParamMap != null) {
+      queryParamMap = HashMultimap.create();
+      for (Map.Entry<String, String> m: simpleQueryParamMap.entrySet()) {
+        queryParamMap.put(m.getKey(), m.getValue());
+      }
+    }
+
+    Multimap<String, String> headerMap = null;
+    if (simpleHeaderMap != null) {
+      headerMap = HashMultimap.create();
+      for (Map.Entry<String, String> h: simpleHeaderMap.entrySet()) {
+        headerMap.put(h.getKey(), h.getValue());
+      }
+    }
+
+    return executeReq(method, region, bucketName, objectName, headerMap, queryParamMap, body, length);
+  }
+
+  private HttpResponse executeReq(Method method, String region, String bucketName, String objectName,
+                               Multimap<String,String> headerMap, Multimap<String,String> queryParamMap,
                                Object body, int length)
     throws InvalidBucketNameException, NoSuchAlgorithmException, InsufficientDataException, IOException,
            InvalidKeyException, NoResponseException, XmlPullParserException, ErrorResponseException,
            InternalException {
     String contentType = null;
-    if (headerMap != null) {
-      contentType = headerMap.get("Content-Type");
+    if (headerMap != null && headerMap.get("Content-Type") != null) {
+      contentType = String.join(" ", headerMap.get("Content-Type"));
```

Comment:
> As this is an `private` method, it would be clean to have `HashMultimap` to avoid `null` value check like here.

I would prefer this additional null check in case when someone does more tweaks in the code later.

Reply:
IMO `private` methods and their usage need to be well controlled.  Its just a thought.